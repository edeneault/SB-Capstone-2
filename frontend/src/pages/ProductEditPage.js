import React, { useState, useEffect } from "react";
import axios from "axios";
import { Container, Form, Button } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import Loader from "../components/Loader";
import FormContainer from "../components/FormContainer";
import { listProductDetails, updateProduct } from "../actions/productActions";
import { PRODUCT_UPDATE_RESET } from "../constants/productConstants";

import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";

toast.configure();

const ProductEditPage = ({ match, history }) => {
  const productId = match.params.id;

  const [name, setName] = useState("");
  const [price, setPrice] = useState(0);
  const [image, setImage] = useState("");
  const [brand, setBrand] = useState("");
  const [category, setCategory] = useState("");
  const [countInStock, setCountInStock] = useState(0);
  const [description, setDescription] = useState("");
  const [uploading, setUploading] = useState(false);
  const [previewSource, setPreviewSource] = useState();

  const dispatch = useDispatch();

  const productDetails = useSelector((state) => state.productDetails);
  const { loading, error, product } = productDetails;

  const productUpdate = useSelector((state) => state.productUpdate);
  const {
    loading: loadingUpdate,
    error: errorUpdate,
    success: successUpdate,
  } = productUpdate;

  const previewFile = (file) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onloadend = () => {
      setPreviewSource(reader.result);
      uploadImage(reader.result);
    };
  };

  useEffect(() => {
    if (successUpdate) {
      dispatch({ type: PRODUCT_UPDATE_RESET });
      history.push("/admin/productslist");
    } else {
      if (!product.name || product._id !== productId) {
        dispatch(listProductDetails(productId));
      } else {
        setName(product.name);
        setPrice(product.price);
        setImage(product.image);
        setBrand(product.brand);
        setCategory(product.category);
        setCountInStock(product.countInStock);
        setDescription(product.description);
      }
    }
  }, [dispatch, history, productId, product, successUpdate]);

  const handleFileInputChange = (e) => {
    const file = e.target.files[0];
    previewFile(file);
  };

  const uploadImage = async (base64EncodedImage) => {
    try {
      setUploading(true);
      const config = {
        headers: {
          "Content-Type": "application/json",
        },
      };

      const body = JSON.stringify({ data: base64EncodedImage });
      await axios
        .post("/api/upload", body, config)
        .then((res) => {
          const secure_url = res.data.secure_url;
          setUploading(false);
          setImage(secure_url);
        })
        .catch((err) => new Error(err));
    } catch (err) {
      setUploading(false);
    }
  };

  const submitHandler = (e) => {
    e.preventDefault();

    dispatch(
      updateProduct({
        _id: productId,
        name,
        price,
        image,
        brand,
        category,
        description,
        countInStock,
      }),
    );
    dispatch(listProductDetails(productId));
  };

  return (
    <>
      <Container fluid className='px-5'>
        <Button className='btn btn-light my-3' onClick={() => history.goBack()}>
          Go Back
        </Button>
        <FormContainer>
          <h1>Edit Product</h1>
          {loadingUpdate && <Loader />}
          {errorUpdate && toast(`${errorUpdate}`, { type: "error" })}
          {loading ? (
            <Loader />
          ) : error ? (
            toast(`${error}`, { type: "error" })
          ) : (
            <Form onSubmit={submitHandler}>
              <Form.Group controlId='name'>
                <Form.Label>Name</Form.Label>
                <Form.Control
                  type='name'
                  placeholder='Enter name'
                  value={name}
                  onChange={(e) => setName(e.target.value)}
                ></Form.Control>
              </Form.Group>

              <Form.Group controlId='price'>
                <Form.Label>Price</Form.Label>
                <Form.Control
                  type='number'
                  placeholder='Enter price'
                  value={price}
                  onChange={(e) => setPrice(e.target.value)}
                ></Form.Control>
              </Form.Group>

              <Form.Group controlId='image'>
                <Form.Label>Image</Form.Label>
                <Form.Control
                  type='text'
                  placeholder='Enter image URL'
                  value={image}
                  onChange={(e) => setImage(e.target.value)}
                ></Form.Control>
                <Form.Control
                  type='file'
                  label='choose file'
                  onChange={handleFileInputChange}
                  className='btn text-dark'
                ></Form.Control>
                {uploading && <Loader />}
              </Form.Group>

              {previewSource && (
                <img
                  src={previewSource}
                  alt='chosen'
                  style={{ height: "200px" }}
                />
              )}

              <Form.Group controlId='brand'>
                <Form.Label>Brand</Form.Label>
                <Form.Control
                  type='text'
                  placeholder='Enter brand'
                  value={brand}
                  onChange={(e) => setBrand(e.target.value)}
                ></Form.Control>
              </Form.Group>

              <Form.Group controlId='countInStock'>
                <Form.Label>Count In Stock</Form.Label>
                <Form.Control
                  type='number'
                  placeholder='Enter count-in-stock'
                  value={countInStock}
                  onChange={(e) => setCountInStock(e.target.value)}
                ></Form.Control>
              </Form.Group>

              <Form.Group controlId='category'>
                <Form.Label>Category</Form.Label>
                <Form.Control
                  type='text'
                  placeholder='Enter category'
                  value={category}
                  onChange={(e) => setCategory(e.target.value)}
                ></Form.Control>
              </Form.Group>

              <Form.Group controlId='description'>
                <Form.Label>Description</Form.Label>
                <Form.Control
                  type='text'
                  placeholder='Enter description'
                  value={description}
                  onChange={(e) => setDescription(e.target.value)}
                ></Form.Control>
              </Form.Group>

              <Button type='submit' variant='primary' className='my-2'>
                Update
              </Button>
            </Form>
          )}
        </FormContainer>
      </Container>
    </>
  );
};

export default ProductEditPage;
