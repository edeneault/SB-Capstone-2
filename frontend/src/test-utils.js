import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

// create a customRender that wraps the UI in a memory Router

const MemoryRouterWithInitialRoutes = ({ children, initialRoutes }) => {
  return <MemoryRouter initialEntries={initialRoutes}>{children}</MemoryRouter>;
};

const customRender = (ui, options) => {
  const initialRoutes =
    options && options.initialRoutes ? options.initialRoutes : ["/"];

  return render(ui, {
    wrapper: (args) =>
      MemoryRouterWithInitialRoutes({
        ...args,
        initialRoutes,
      }),
    ...options,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
