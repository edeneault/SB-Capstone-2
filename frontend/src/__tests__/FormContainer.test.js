import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { render } from "../test-utils";
import { initialState } from "../store-mock";
import FormContainer from "../components/FormContainer";
import ProductEditPage from "../pages/ProductEditPage";

const mockStore = configureMockStore([thunk]);

describe("test", () => {
  test("renders FormContainer with children without crashing", () => {
    const store = mockStore(initialState);
    const match = {
      params: {
        id: "61306f7d7ef9b1460b0f6c6c",
      },
    };

    const { getByText } = render(
      <Provider store={store}>
        <FormContainer children={<ProductEditPage match={match} />} />
      </Provider>,
    );

    expect(getByText("Edit Product")).toBeInTheDocument();
  });

  test("matches snapshot", function () {
    const store = mockStore(initialState);
    const match = {
      params: {
        id: "61306f7d7ef9b1460b0f6c6c",
      },
    };

    const { asFragment } = render(
      <Provider store={store}>
        <FormContainer children={<ProductEditPage match={match} />} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
