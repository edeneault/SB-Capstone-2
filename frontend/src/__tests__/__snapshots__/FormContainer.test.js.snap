// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="justify-content-md-center row"
    >
      <div
        class="col-md-6 col-12"
      >
        <div
          class="px-5 container-fluid"
        >
          <button
            class="btn btn-light my-3 btn btn-primary"
            type="button"
          >
            Go Back
          </button>
          <div
            class="container"
          >
            <div
              class="justify-content-md-center row"
            >
              <div
                class="col-md-6 col-12"
              >
                <h1>
                  Edit Product
                </h1>
                <form
                  class=""
                >
                  <div
                    class="form-group"
                  >
                    <label
                      class="form-label"
                      for="name"
                    >
                      Name
                    </label>
                    <input
                      class="form-control"
                      id="name"
                      placeholder="Enter name"
                      type="name"
                      value=""
                    />
                  </div>
                  <div
                    class="form-group"
                  >
                    <label
                      class="form-label"
                      for="price"
                    >
                      Price
                    </label>
                    <input
                      class="form-control"
                      id="price"
                      placeholder="Enter price"
                      type="number"
                      value="0"
                    />
                  </div>
                  <div
                    class="form-group"
                  >
                    <label
                      class="form-label"
                      for="image"
                    >
                      Image
                    </label>
                    <input
                      class="form-control"
                      id="image"
                      placeholder="Enter image URL"
                      type="text"
                      value=""
                    />
                    <input
                      class="btn text-dark form-control-file"
                      id="image"
                      label="choose file"
                      type="file"
                    />
                  </div>
                  <div
                    class="form-group"
                  >
                    <label
                      class="form-label"
                      for="brand"
                    >
                      Brand
                    </label>
                    <input
                      class="form-control"
                      id="brand"
                      placeholder="Enter brand"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="form-group"
                  >
                    <label
                      class="form-label"
                      for="countInStock"
                    >
                      Count In Stock
                    </label>
                    <input
                      class="form-control"
                      id="countInStock"
                      placeholder="Enter count-in-stock"
                      type="number"
                      value="0"
                    />
                  </div>
                  <div
                    class="form-group"
                  >
                    <label
                      class="form-label"
                      for="category"
                    >
                      Category
                    </label>
                    <input
                      class="form-control"
                      id="category"
                      placeholder="Enter category"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="form-group"
                  >
                    <label
                      class="form-label"
                      for="description"
                    >
                      Description
                    </label>
                    <input
                      class="form-control"
                      id="description"
                      placeholder="Enter description"
                      type="text"
                      value=""
                    />
                  </div>
                  <button
                    class="my-2 btn btn-primary"
                    type="submit"
                  >
                    Update
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
